charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,249 @@
-                            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
-                            "BACKEND_MODE": config["backend-mode"],
-                        },
-                        "ports": [{"name": "http", "containerPort": config["port"]}],
-                    }
-                ],
             },
-        )
+        }
 
-        self.model.unit.status = ActiveStatus()
+        return Layer(layer_config)
 
-    def _configure_mesh(self, interfaces):
+    def _on_install(self, _) -> None:
+        """Perform installation only actions."""
+        try:
+            # Deploy K8S resources to speed up deployment
+            self._deploy_k8s_resources()
+        except ErrorWithStatus as error:
+            self._log_and_set_status(error.status)
+        return
+
+    def _on_remove(self, _) -> None:
+        """Remove all resources."""
+        self._log_and_set_status(MaintenanceStatus("Removing K8S resources"))
+        try:
+            self.k8s_resource_handler.delete()
+        except ApiError as error:
+            # Do not log/report when resources were not found
+            if error.status.code != 404:
+                self.logger.error(f"Removing K8S resources failed with error: {error}")
+                raise GenericCharmRuntimeError("Removing K8s resources failed") from error
+
+    def _on_event(self, event) -> None:
+        """Perform required actions for every event."""
+        try:
+            self._check_leader()
+            self._deploy_k8s_resources()
+            update_layer(
+                self._container_name,
+                self._container,
+                self._tensorboards_web_app_layer,
+                self.logger,
+            )
+            interfaces = self._get_interfaces()
+            self._configure_mesh(interfaces)
+            self.unit.status = ActiveStatus()
+
+        except ErrorWithStatus as error:
+            self._log_and_set_status(error.status)
+            return
+
+        self.unit.status = ActiveStatus()
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploy K8S resources."""
+        try:
+            self._log_and_set_status(MaintenanceStatus("Creating K8S resources"))
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("Creating K8S resources failed", BlockedStatus)
+        self._log_and_set_status(MaintenanceStatus("K8S resources created"))
+
+    def _configure_mesh(self, interfaces) -> None:
         if interfaces["ingress"]:
             interfaces["ingress"].send_data(
                 {
                     "prefix": "/tensorboards",
                     "rewrite": "/",
-                    "service": self.model.app.name,
-                    "port": self.model.config["port"],
+                    "service": self.app.name,
+                    "port": PORT,
                 }
             )
+        else:
+            raise ErrorWithStatus("No ingress relation available", BlockedStatus)
```

Comment:
The team has agreed the BlockedStatus messages should give instructions to resolve, in this case something like `"Please relate to <relation>"`